src/tactic/fix_by_cases.lean

```diff
@@ -0,0 +1,32 @@
+/-
+Copyright (c) 2020 Gabriel Ebner. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Gabriel Ebner
+-/
+
+/-!
+# Fix the `by_cases` tactic
+
+The core `by_cases` tactic has several bugs:
+ - It only works if the proposition is decidable.
+ - It sometimes unfolds the proposition.
+
+We override the `by_cases` tactic with a correct implementation here.
+-/
+
+namespace tactic
+
+/--
+Do not use this function directly, use `tactic.by_cases`.
+-/
+meta def by_cases' (e : expr) (h : name) : tactic unit := do
+dec_e ← mk_app ``decidable [e] <|> fail "by_cases tactic failed, type is not a proposition",
+inst ← mk_instance dec_e <|> pure `(classical.prop_decidable %%e),
+let inst := `(id %%inst : decidable %%e),
+cases inst [h, h],
```

Comment:
Does anybody use `by_cases` in term-mode and expect the result to be a `dite`?

Reply:
Apparently @b-mehta's `fintype_hom` instance did, so I've changed it (back) to produce a `dite` term.

Reply:
Is this something I can fix? Was it one of the things in `wide_pullbacks`?

Reply:
No, I've already changed `by_cases` so that it produces a `dite`.  I don't have a strong opinion one way or another.  The `cases` tactic (which produces `decidable.rec`) was just easier to use than to construct a `dite` term by hand.

And yes, it was `wide_pullbacks`.

The only thing missing on this PR now are the linting errors (inhabited instances & docstrings).